pkg/roles/service.go

```diff
@@ -77,56 +77,66 @@ func (s *Service) GetRole(ctx context.Context, userID, ID string) (*Role, error)
 	ctx, span := s.tracer.Start(ctx, "roles.Service.GetRole")
 	defer span.End()
 
-	exists, err := s.ofga.Check(ctx, fmt.Sprintf("user:%s", userID), "can_view", fmt.Sprintf("role:%s", ID))
+	role, err := s.repo.FindRoleByIdAndOwner(ctx, ID, userID)
+
+	if errors.Is(err, storage.ErrNotFound) {
+		s.logger.Debugf("role %s for owner %s not found", ID, userID)
+		return nil, nil
+	}
 
 	if err != nil {
+		err = fmt.Errorf("unable to get role %s for owner %s, %w", ID, userID, err)
 		s.logger.Error(err.Error())
 		return nil, err
 	}
 
-	if !exists {
-		return nil, nil
-	}
-
-	role := new(Role)
-	role.ID = ID
-	role.Name = ID
-
 	return role, nil
 }
 
 // CreateRole creates a role and associates it with the userID passed as argument
-// an extra tuple is created to estabilish the "privileged" relatin for admin users
-func (s *Service) CreateRole(ctx context.Context, userID, ID string) (*Role, error) {
+// an extra tuple is created to estabilish the "privileged" relation for admin users
+func (s *Service) CreateRole(ctx context.Context, userID, roleName string) (*Role, error) {
 	ctx, span := s.tracer.Start(ctx, "roles.Service.CreateRole")
 	defer span.End()
 
-	// TODO @shipperizer @barco will we need also the can_edit, can_delete?
-	// does creating a role mean that you are the owner, therefore u get all the permissions on it?
-	// right now assumption is only admins will be able to do this
-	// potentially changing the model to say
-	// `define can_view: [user, user:*, role#assignee, group#member] or assignee or admin from privileged`
-	// might sort the problem
+	createdRole, tx, err := s.repo.CreateRoleTx(ctx, userID, roleName)
+	if err != nil {
+		err = fmt.Errorf("unable to create role %s for user %s, %w", roleName, userID, err)
+		s.logger.Error(err.Error())
+		return nil, err
+	}
 
-	// TODO @shipperizer offload to privileged creator object
-	role := fmt.Sprintf("role:%s", ID)
+	role := fmt.Sprintf("role:%s", roleName)
```

Comment:
isn't it better sticking with the id in OpenFGA too?see the complication of not having the id until that is returned by the dB, but wondering if this is wise in light of competing names (even though we should have a unique index on name?)

Reply:
Synced offline